File: FlashCards.py
#!/usr/bin/python

# Author: Chris Robertson https://github.com/electronicsleep
# Date: 09/10/2014
# Purpose: Simple FlashCard app for learning math and python
# Released under the BSD license
# Works with python 2 and 3

# Running:
# python FlashCards.py

# IMPORT LIBRARIES

import time
import datetime
import sys

from random import randint
from sys import argv

# CARD FUNCTIONS


def show_question_card(card_front):
    print("\---------------------------/")
    print("|///////////////////////////|")
    print("|                           |")
    print("| Q. %s                " % card_front.strip())
    print("|                           |")
    print("|///////////////////////////|")
    print("\---------------------------/")


def show_answer_card(card_back):
    print("\---------------------------/")
    print("|                           |")
    print("|                           |")
    print("| A. %s               " % card_back.strip())
    print("|                           |")
    print("|                           |")
    print("\---------======------------/")


def main():

    # PRINT NUMBER OF ARGUMENTS
    # print(len(sys.argv))

    if len(sys.argv) == 2:
        script, username = argv
    else:
        script = argv
        username = "Emtpy"

    print("Script:", script)
    print("User: ", username)

    time_start = time.time()

    prompt = '> '
    card_front = ""

    # SETUP CARD DICTIONARY

    # INITIAL SETUP
    cards = {}
    mem_file = ""
    question_line = ""

    # IMPORT MORE CARDS FROM FILE
    try:
        mem_file = open('memorize.txt', 'r')
    except Exception as e:
        print("Please create a memorize.txt file")
        print("Error: " + str(e))
        exit()

    # PRINT ENTIRE FILE OF QA CARDS
    # print mem_file.read()

    # PARSE FILE FOR QUESTION AND ANSWER FOR CARDS

    for line in mem_file:
        if line.startswith('Q.'):
            first, _, question_line = line.partition(" ")
            # print("LOAD: Question: " + line,)
        elif line.startswith('A.'):
            first, _, answer_line = line.partition(" ")
            cards[question_line] = answer_line
            # print("LOAD: Answer: " + line,)

    mem_file.close()

    # FIND HOW MANY CARDS

    n = len(cards.keys())
    print("\=====-----======--------=====/")
    print("NUMBER OF CARDS %s" % n)
    print("DATE: " + datetime.datetime.now().strftime("%m/%d/%Y %H:%M:%S"))

    num_cards = 4

    num_correct = 0
    num_incorrect = 0

    for x in range(num_cards):

        random_num = randint(1, n)

        # LOOP AND UNPACK ONE RANDOM CARD
        i = 0
        for k, v in cards.items():

            i += 1
            if i == random_num:
                card_front = k
                card_back = v

        show_question_card(card_front)

        print("What is the answer? ")

        answer = input(prompt)
        answer_str = str(answer)
        answer_str = answer_str.lower()
        answer_card = card_back.lower()

        if answer_str.strip() == answer_card.strip():
            print("*** CORRECT ***")
            num_correct += 1
            print("Num Correct: ", num_correct)
        else:
            print("*** INCORRECT ***")
            print("The correct answer is:")
            print("the answer is: |")
            print(card_back.strip())
            print("|")
            num_incorrect += 1
            print("Num Incorrect: ", num_incorrect)

        show_answer_card(card_back)

        # GET END TIME
        time_end = time.time()

        # PRINT TIME IT TAKES TO ANSWER
        diff = time_end - time_start
        diff_str = str(diff)
        print("DATE: " + datetime.datetime.now().strftime("%m/%d/%Y %H:%M:%S"))
        print("You took " + diff_str + " seconds to answer")
        percent_correct = ((float(num_correct) / float(num_cards)) * 100)
        if percent_correct > 75:
            print ("Good Job!")
        print(percent_correct, "% Correct")


if __name__ == "__main__":
    main()

File: FlashCards.cpp
/*
Author: Chris Robertson <electronicsleep@gmail.com>
Date: 12/11/2016
Purpose: Simple FlashCard app for learning math and C++
Released under the BSD license

Running:
cppcheck FlashCards.cpp
llvm-g++ FlashCards.cpp -o ./FlashCards
g++ FlashCards.cpp -o FlashCards
./FlashCards
*/

#include <iostream>
#include <fstream>
#include <cstring>
#include <vector>
#include <unistd.h>
#include <stdlib.h>

using namespace std;

bool does_file_exist(string fileName)
{
  std::ifstream infile(fileName);
  return infile.good();
}


int main(int argc, char** argv)
{

  //Number of FlashCards to run
  int total_drill_num = 4;

  string mem_file = "memorize.txt";
  string answer = "";
  string text = "";

  if (!does_file_exist(mem_file))
  {
    cout << mem_file + " file does not exist" << endl;
    return 1;
  } else {
    cout << "Using Memory File: " + mem_file << endl;

    ifstream ifs(mem_file);

    int num = 0;
    int card = 0;

    std::vector<std::string> answers;
    std::vector<std::string> questions;

    while(!ifs.eof()) 
    {
      getline(ifs, text);

      if (text.find("Q.") != std::string::npos) {
        questions.push_back(string(text));
      } else if (text.find("A.") != std::string::npos) {
        answers.push_back(string(text));
        num++;
      }

    }

    if (num == 0)
    {
      cout << "No cards in " + mem_file + " file" << endl;
      return 1;
    } 

    cout << "Number of Cards: " << num << endl;

 
    int drill_num = 0;
    int num_correct = 0;
    int num_incorrect = 0;

    for (drill_num = 0; drill_num < total_drill_num; drill_num++)
    {
      //Get random card
      srand(time(NULL));
      card = rand() % num; 
 
      //Show random card
      cout << "/=======-----===========-----=======/" << endl;
      cout << "/=======-----Random Card-----=======/" << endl;
      cout << "Question: " << questions[card] << endl;
      //usleep(3000000); 

      cout << "Enter Answer: "; 
      cin >> answer;

      //cout << answers[card] << endl;
      //cout << answer << endl;

      if (answers[card] == ("A. " + answer)) 
      {
        cout << "Correct" << endl;
        num_correct++;
      } else {
        cout << "Incorrect" << endl;
        num_incorrect++;
      }

      cout << "Answer: " << answers[card] << endl;
      cout << "-----------------------------------=/" << endl;
      cout << " -----------===========-----------=//" << endl;
      printf("Correct: %d Incorrect: %d Card: %d\n\n", num_correct, num_incorrect, drill_num + 1);
      float ans = ((float)num_correct/ ((float)drill_num + 1) * 100);
      printf("%f Correct\n", ans);
    }

  }
  return 0;
}
